### PR TITLE
Adding segmentation output to Stable Diffusion pipeline

### PR DESCRIPTION
**Pull request by orpatashnik**
Initial commit for adding a pipeline that outputs a segmentation map in addition to the generated image.  The segmentation is done with the "self-segmentation" method presented here https://arxiv.org/abs/2303.11306

### CODE REVIEW

**Comment by sayakpaul** [comment]
@orpatashnik thanks so much for opening the PR. 

I think we still need to add a couple more files, but let us know if you'd like an initial review.

File: src/diffusers/pipelines/stable_diffusion_self_segment_pipeline/requirements.txt [hunk 1]
```diff
@@ -0,0 +1,3 @@
+transformers
```
Comment (patrickvonplaten) [inline src/diffusers/pipelines/stable_diffusion_self_segment_pipeline/requirements.txt:1]: We cannot add these dependencies in a requirement file, let's try to make the optional

File: src/diffusers/pipelines/stable_diffusion_self_segment_pipeline/README.md [hunk 2]
```diff
@@ -0,0 +1,132 @@
+# Self-Segmentation for Stable Diffusion
```
Comment (patrickvonplaten) [inline src/diffusers/pipelines/stable_diffusion_self_segment_pipeline/README.md:1]: Let's try to move this in the docs :-)

File: src/diffusers/pipelines/stable_diffusion_self_segment_pipeline/stable_diffusion_self_segment.py [hunk 3]
```diff
@@ -0,0 +1,963 @@
...
+import warnings
+from typing import Any, Callable, Dict, List, Optional, Union
+
+import nltk
+import numpy as np
+import torch
+from diffusers.models.attention_processor import Attention
+from diffusers.pipelines.stable_diffusion import StableDiffusionSafetyChecker
+from packaging import version
+from sklearn.cluster import SpectralClustering, KMeans
```
Comment (patrickvonplaten) [inline src/diffusers/pipelines/stable_diffusion_self_segment_pipeline/stable_diffusion_self_segment.py:25]:
```suggestion
if is_sklearn_available():
    from sklearn.cluster import SpectralClustering, KMeans
```

File: src/diffusers/pipelines/stable_diffusion_self_segment_pipeline/stable_diffusion_self_segment.py [hunk 4]
```diff
@@ -0,0 +1,963 @@
...
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import inspect
+import warnings
+from typing import Any, Callable, Dict, List, Optional, Union
+
+import nltk
```
Comment (patrickvonplaten) [inline src/diffusers/pipelines/stable_diffusion_self_segment_pipeline/stable_diffusion_self_segment.py:19]:
```suggestion
if is_nltk_available():
    import nltk
```

File: src/diffusers/pipelines/self_segmentation_stable_diffusion/self_segmentation_stable_diffusion.py [hunk 5]
```diff
@@ -0,0 +1,968 @@
+# Copyright 2023 The HuggingFace Team. All rights reserved.
```
Comment (sayakpaul) [inline src/diffusers/pipelines/self_segmentation_stable_diffusion/self_segmentation_stable_diffusion.py:1]:
```suggestion
# Copyright 2023 Self Segmentation Stable Diffusion Authors and The HuggingFace Team. All rights reserved.
```

File: src/diffusers/pipelines/self_segmentation_stable_diffusion/self_segmentation_stable_diffusion.py [hunk 6]
```diff
@@ -0,0 +1,968 @@
...
+
+logger = logging.get_logger(__name__)  # pylint: disable=invalid-name
+
+EXAMPLE_DOC_STRING = """
+    Examples:
+        ```py
+        >>> import torch
+        >>> from diffusers import StableDiffusionPipeline
+
+        >>> pipe = StableDiffusionPipeline.from_pretrained("runwayml/stable-diffusion-v1-5", torch_dtype=torch.float16)
```
Comment (sayakpaul) [inline src/diffusers/pipelines/self_segmentation_stable_diffusion/self_segmentation_stable_diffusion.py:60]: Shouldn't this be `SelfSegmentationStableDiffusionPipeline`?

File: src/diffusers/pipelines/self_segmentation_stable_diffusion/self_segmentation_stable_diffusion.py [hunk 7]
```diff
@@ -0,0 +1,968 @@
...
+                that correspond to the same object.
+            seg_labels_indices_in_prompt (`List[int]`, defaults to None):
+                The indices of words in the prompt that corresponds to object labels. The labeling algorithm look at
+                the cross-attention maps of these indices. If None, nouns are automatically extracted from the prompt
+                and their indices are used. Automatically extracting the nouns requires nltk package.
+
+        Examples:
+
+        Returns:
+            [`~pipelines.stable_diffusion.BaseOutput`] or `tuple`: [`~pipelines.stable_diffusion.BaseOutput`] if
```
Comment (sayakpaul) [inline src/diffusers/pipelines/self_segmentation_stable_diffusion/self_segmentation_stable_diffusion.py:823]: Could we try to format this like 

https://github.com/huggingface/diffusers/blob/7a91ea6c2b53f94da930a61ed571364022b21044/src/diffusers/pipelines/stable_diffusion_xl/pipeline_stable_diffusion_xl.py#L648

?

File: src/diffusers/pipelines/self_segmentation_stable_diffusion/self_segmentation_stable_diffusion.py [hunk 8]
```diff
@@ -0,0 +1,968 @@
...
+            prompt_embeds.dtype,
+            device,
+            generator,
+            latents,
+        )
+
+        # 6. Prepare extra step kwargs. TODO: Logic should ideally just be moved out of the pipeline
+        extra_step_kwargs = self.prepare_extra_step_kwargs(generator, eta)
+
+        self.attn_processor.reset()
```
Comment (sayakpaul) [inline src/diffusers/pipelines/self_segmentation_stable_diffusion/self_segmentation_stable_diffusion.py:890]: Add a comment here?

File: src/diffusers/pipelines/self_segmentation_stable_diffusion/self_segmentation_stable_diffusion.py [hunk 9]
```diff
@@ -0,0 +1,968 @@
...
+        image = self.image_processor.postprocess(image, output_type=output_type, do_denormalize=do_denormalize)
+
+        # Offload last model to CPU
+        if hasattr(self, "final_offload_hook") and self.final_offload_hook is not None:
+            self.final_offload_hook.offload()
+
+        if not return_dict:
+            return image, has_nsfw_concept, segmentation_map, seg_labels, merged_seg_map, new_labels
+
+        return BaseOutput(
```
Comment (sayakpaul) [inline src/diffusers/pipelines/self_segmentation_stable_diffusion/self_segmentation_stable_diffusion.py:961]: It would make sense to write `SelfSegmentationStableDiffusionPipelineOutput` for packing the outputs.

**Review by sayakpaul — COMMENTED** [review]
Cool, this looks great to me.

Lets add tests to ship this 🚀

**Comment by sayakpaul** [comment]
@orpatashnik, additionally, I have:

* Improved the documentation a bit. 
* Resolved the conflicts.

File: src/diffusers/pipelines/self_segmentation_stable_diffusion/self_segmentation_stable_diffusion.py (hunk 7 above)
Comment (orpatashnik) [reply src/diffusers/pipelines/self_segmentation_stable_diffusion/self_segmentation_stable_diffusion.py]: @sayakpaul I am not sure I understood you correctly, but if I did then I fixed it

File: docs/source/en/_toctree.yml [hunk 10]
```diff
@@ -224,6 +224,8 @@
       title: Score SDE VE
     - local: api/pipelines/self_attention_guidance
       title: Self-Attention Guidance
+    - local: api/pipelines/self-segmentation-stable-diffusion
```
Comment (sayakpaul) [inline docs/source/en/_toctree.yml:227]: The name should have been `self_segmentation_stable_diffusion`.

File: src/diffusers/pipelines/self_segmentation_stable_diffusion/self_segmentation_stable_diffusion.py [hunk 11]
```diff
@@ -0,0 +1,1008 @@
...
+    import nltk
+
+
+logger = logging.get_logger(__name__)  # pylint: disable=invalid-name
+
+EXAMPLE_DOC_STRING = """
+    Examples:
+        ```py
+        >>> import torch
+        >>> from diffusers import StableDiffusionPipeline
```
Comment (sayakpaul) [inline src/diffusers/pipelines/self_segmentation_stable_diffusion/self_segmentation_stable_diffusion.py:58]: Should have been `SelfSegmentationStableDiffusionPipeline`.

**Comment by sayakpaul** [comment]
@orpatashnik let's try to fix the quality related checks on the CI by running

* `make style && make quality`. 
* `make fix-copies`.

These are well documented here: https://github.com/huggingface/diffusers/blob/main/CONTRIBUTING.md. 

Fixing this should hopefully also fix the documentation problem. 

Also, there seems to be an error in the test script, which is why the related tests on the CLI is failing. To ensure the tests are passing locally, you can do:

```
RUN_SLOW=1 pytests <path_to_the_test_script>
```

Let us know if anything is unclear.
